core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -100,6 +106,53 @@ public static EnumerableHashJoin create(
         condition, variablesSet, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null || collation == RelCollations.EMPTY) {
+      return null;
+    }
+
+    List<Integer> requiredKeys = RelCollations.ordinals(collation);
+    ImmutableBitSet requiredKeySet = ImmutableBitSet.of(requiredKeys);
+
+    ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
+    ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
+        .shift(left.getRowType().getFieldCount());
+
+    // HashJoin traits passdown shall only consider left/right outer join.
+    // It is because for a hash-based implementation, only non-hashed side can
+    // preserve ordering, thus only for left/right outer join, we are sure which
+    // side is non-hashed side (the outer child).
+    if (joinType == JoinRelType.LEFT) {
```

Comment:
INNER join, SEMI/ANTI SEMI, LEFT all can pass request to left child, and derive from left child.

Reply:
Only right outer join uses right child as the probe side, I guess. Need to check the implementation of HashJoin.

Reply:
On the other hand, it shows exactly that trait propagation should only be done on physical operators, because it is physical implementation dependent.

Reply:
Agreed. I will check EnumerableHashJoin. Basically collation passes down will depend on which side to build the hash table.

Reply:
Please note that currently EnumerableHashJoin implementation builds the hash table always on the right input.

Reply:
Thanks @rubenada for the reminder. It's surprising though. Let me check the implementation of EnumerableHashJoin to see how it make right to build hash table on right input for **RIGHT OUTER JOIN**.

Reply:
Then for FULL/RIGHT OUTER JOIN, we should prohibit trait propagations.